kubernetes/kubernetes.go

```diff
@@ -413,6 +440,79 @@ func (ka KubeAgentConfig) sendMetrics(metricSample *os.File) {
 	}
 }
 
+func (ka KubeAgentConfig) sendMetricsBasedOnUploadMode(customS3Mode bool, metricSample *os.File) {
+	if customS3Mode {
+		log.Infof("Uploading Metrics to Custom S3 Bucket %s", ka.CustomS3UploadBucket)
+		go ka.sendMetricsToCustomS3(metricSample)
+	} else {
+		log.Info("Uploading Metrics")
+		go ka.sendMetrics(metricSample)
+	}
+}
+
+func (ka KubeAgentConfig) sendMetricsToCustomS3(metricSample *os.File) {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(ka.CustomS3Region)},
+	)
+	if err != nil {
+		log.Fatalf("Could not establish AWS Session, "+
+			"ensure AWS environment variables are set correctly: %s", err)
+	}
+	svc := s3.New(sess)
+
+	key := generateSampleKey(ka.clusterUID)
+
+	metricFile, err := os.Open(metricSample.Name())
+	if err != nil {
+		log.Fatalf("Failed to open metric sample file: %v", err)
+	}
+	defer metricFile.Close()
+
+	fileInfo, err := metricFile.Stat()
+	if err != nil {
+		log.Fatalf("Failed to get metric sample file statistics: %v", err)
+	}
+	fileSize := fileInfo.Size()
+	fileBuffer := make([]byte, fileSize)
+	_, err = metricFile.Read(fileBuffer)
+	if err != nil {
+		log.Fatalf("Failed to read metric sample file: %v", err)
+	}
+
+	_, err = svc.PutObject(&s3.PutObjectInput{
+		Bucket:        aws.String(ka.CustomS3UploadBucket),
+		Key:           aws.String(key),
+		Body:          bytes.NewReader(fileBuffer),
+		ContentLength: aws.Int64(fileSize),
+		ContentType:   aws.String(http.DetectContentType(fileBuffer))})
+	if err != nil {
+		log.Fatalf("Failed to put Object to custom S3 with error: %s", err)
+	} else {
+		sn := strings.Split(metricSample.Name(), "/")
+		log.Infof("Exported metric sample %s to custom S3 bucket: %s",
+			strings.TrimSuffix(sn[len(sn)-1], ".tgz"), ka.CustomS3UploadBucket)
+	}
+	err = os.Remove(metricSample.Name())
+	if err != nil {
+		log.Warnf("Warning: Unable to cleanup after metric sample upload: %v", err)
+	}
+}
+
+// generateSampleKey creates a key (location) for s3 to upload the sample to. Example of s3 location format
+// <S3_BUCKET>/production/data/metrics-agent/<YYYY>/<MM>/<DD>/<CLUSTER_UID>/<CLUSTER_UID>-<YYYYMMDD>-<HH>-<MM>.tgz
+func generateSampleKey(clusterUID string) string {
+	fileKey := "/production/data/metrics-agent/"
+	currentTime := time.Now()
+	year, month, day := currentTime.Date()
+	hour := currentTime.Hour()
+	minute := currentTime.Minute()
+	fileKey += strconv.Itoa(year) + "/" + fmt.Sprintf("%02d", int(month)) + "/" +
```

Comment:
Utilizing `fmt.Sprintf(....)` would be a lot more clear here.

Reply:
Yup, I agree. Updated